---
title: "Figure 02b - Panel C"
output: 
  rmdformats::robobook:
    lightbox: true
    fig_caption: true
---

<style type="text/css">
  body{
  font-size: 10pt;
}
</style>

# 1) Setup

## - Defining work directory

In this section, we define the working directory for the R Markdown document.

```{r setup, include=TRUE}
# Get the path of the current script
# Then get the parent directory of the parent directory of the parent directory
local_wd_folder <- dirname(dirname(dirname(rstudioapi::getSourceEditorContext()$path)))

# Set the root directory for knitr to the local working directory
knitr::opts_knit$set(root.dir = local_wd_folder)
```

## - Defining input data and output directories

Here, we define the directories for input data and output files.

```{r}
# Get the directory of the current script
script_folder <- dirname(rstudioapi::getSourceEditorContext()$path)

# Define the data folder and output folder
data_folder <- './Data'
#results_folder <- './Results'
figures_folder <- './Figures'
```

## - Setting seed

Setting a seed ensures that any random processes are reproducible.

```{r}
# Set a seed for reproducibility
set.seed(123)
```

## - Packages installation (optional)

In this section, we ensure that all necessary packages are installed.

```{r}
# Ensure BiocManager is available for installation of Bioconductor packages
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(version = "3.18")

# Define a list of required packages used in this script
packages_required <- c("escape")

# Identify any required packages that are not installed
packages_uninstalled <- packages_required[!(packages_required %in% installed.packages()[,"Package"])]

# Install any uninstalled packages
if(length(packages_uninstalled)) BiocManager::install(packages_uninstalled)
```

## - Loading packages

Here, we load the necessary packages for our analysis.

```{r message=FALSE, warning=FALSE}
# Load stringr for string manipulation
library(escape, quietly = TRUE)
```

## - Log Session Info

Finally, we log the session information for reproducibility.

```{r message=TRUE, warning=TRUE}
# Write the session information to a text file
writeLines(capture.output(sessionInfo()), file.path(script_folder, 'Figure_02b_SessionInfo.txt'))

# Print the session information
sessionInfo()
```

# 2) Loading input files

```{r}
Expression_matrix <- read.csv(paste(data_folder, '/RNASeq/Counts/Processed_counts/Grouped/Tximport_TPM_mat.csv', 
                                    sep = ''), 
                              row.names = 'X')

colnames(Expression_matrix) <- str_replace_all(colnames(Expression_matrix), pattern = 'X', replacement = '')

Genes_annotation_metadata <- read.csv(paste(data_folder, "/RNASeq/Annotation_files/GENCODE_v42.gene_metadata.csv", sep = ''))
rownames(Genes_annotation_metadata) <- Genes_annotation_metadata$gene_id

Samples_metadata <- read.csv(paste(data_folder, '/RNASeq/Metadata/RNASeq_Samples_Metadata.csv', sep = ''), row.names = 'Patient.ID')
```

# 3) Ordering expression matrix (TPM)

```{r}
Unique_genes <- read.table(file = paste(data_folder, '/RNASeq/Counts/List_unique_genes.csv', 
                                    sep = ''), header = T, sep = ',')$unique_genes

Expression_matrix <- Expression_matrix[Unique_genes, ]

rownames(Expression_matrix) <- Genes_annotation_metadata[rownames(Expression_matrix),'gene_name']

Expression_matrix <- as.matrix(Expression_matrix)

Expression_matrix <- Expression_matrix[,rownames(Samples_metadata)]
```

# 4) Creating folders for Pathway Analyses

```{r}
dir.create(file.path(paste(data_folder, '/RNASeq/Pathway_Analysis', 
                                    sep = '')), showWarnings = FALSE)

dir.create(file.path(paste(data_folder, '/RNASeq/Pathway_Analysis/ssEnrichments', 
                                    sep = '')), showWarnings = FALSE)

dir.create(file.path(paste(data_folder, '/RNASeq/Pathway_Analysis/ssEnrichments/CellType', 
                                    sep = '')), showWarnings = FALSE)

dir.create(file.path(paste(data_folder, '/RNASeq/Pathway_Analysis/ssEnrichments/Drug_Resistance', 
                                    sep = '')), showWarnings = FALSE)

dir.create(file.path(paste(data_folder, '/RNASeq/Pathway_Analysis/ssEnrichments/Drug_Response', 
                                    sep = '')), showWarnings = FALSE)

dir.create(file.path(paste(data_folder, '/RNASeq/Pathway_Analysis/Control_vs_BP-CMML', 
                                    sep = '')), showWarnings = FALSE)
```

# 5) Gene signature

## - CellTypes

### - 2019 VanGalen et al AML signatures

```{r}
VanGalen_CellType <- read.csv(paste(data_folder, '/RNASeq/Resources/Gene_Pathways/2019__VanGalen_et_al__AML_signatures.csv', 
                                    sep = ''))

HSC.like <- GSEABase::GeneSet(VanGalen_CellType$HSC.like)
HSC.like <- GSEABase::`setName<-`(object = HSC.like, value='HSC.like')

Progenitor.like <- GSEABase::GeneSet(VanGalen_CellType$Progenitor.like)
Progenitor.like <- GSEABase::`setName<-`(object = Progenitor.like, value='Progenitor.like')

HSC.Prog.like <- GSEABase::GeneSet(VanGalen_CellType$HSC.Prog.like)
HSC.Prog.like <- GSEABase::`setName<-`(object = HSC.Prog.like, value='HSC.Prog.like')

GMP.like <- GSEABase::GeneSet(VanGalen_CellType$GMP.like)
GMP.like <- GSEABase::`setName<-`(object = GMP.like, value='GMP.like')

Promono.like <- GSEABase::GeneSet(VanGalen_CellType$Promono.like)
Promono.like <- GSEABase::`setName<-`(object = Promono.like, value='Promono.like')

Monocyte.like <- GSEABase::GeneSet(VanGalen_CellType$Monocyte.like)
Monocyte.like <- GSEABase::`setName<-`(object = Monocyte.like, value='Monocyte.like')

cDC.like <- GSEABase::GeneSet(VanGalen_CellType$cDC.like)
cDC.like <- GSEABase::`setName<-`(object = cDC.like, value='cDC.like')

Myeloid.like <- GSEABase::GeneSet(VanGalen_CellType$Myeloid.like)
Myeloid.like <- GSEABase::`setName<-`(object = Myeloid.like, value='Myeloid.like')

VanGalen_CellType_collection <- GSEABase::GeneSetCollection(HSC.like, HSC.Prog.like, Progenitor.like, 
                                                            GMP.like, Promono.like, Monocyte.like, 
                                                            cDC.like, Myeloid.like)

# - - - - without healthies

VanGalen_CellType_scaled_results_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,rownames(Samples_metadata[Samples_metadata$Condition == 'BP-CMML',])]), 
               gene.sets = VanGalen_CellType_collection, 
               cores = 12, 
               ssGSEA.norm = T,
               min.size = 5)

write.csv(VanGalen_CellType_scaled_results_no_healthies, paste(data_folder, '/RNASeq/Pathway_Analysis/ssEnrichments/CellType/2019__VanGalen_et_al__AML_signatures__Enrichments.csv', 
                                    sep = ''),
          quote = F, row.names = T)
```
### - 2017 Velten et al Healthy signatures

```{r}
Velten_CellType_Modules <- read.csv(paste(data_folder, '/RNASeq/Resources/Gene_Pathways/2017__Velten_et_al__Healthy_signatures.csv', 
                                    sep = ''))

library(stringi)

Hematopoietic_module_genes <- list()

for (Hematopoietic_module in colnames(Velten_CellType_Modules)) {

Hematopoietic_module_genes[[Hematopoietic_module]] <-  Velten_CellType_Modules[[Hematopoietic_module]][Velten_CellType_Modules[[Hematopoietic_module]] != ""]
  
Hematopoietic_module_genes[[Hematopoietic_module]] <- str_split_fixed(Hematopoietic_module_genes[[Hematopoietic_module]], pattern = ' ', n = 2)[,1]

Hematopoietic_module_genes[[Hematopoietic_module]] <- as.vector(na.omit(Hematopoietic_module_genes[[Hematopoietic_module]]))

Hematopoietic_module_genes[[Hematopoietic_module]] <- GSEABase::GeneSet(Hematopoietic_module_genes[[Hematopoietic_module]])

Hematopoietic_module_genes[[Hematopoietic_module]] <- GSEABase::`setName<-`(object = Hematopoietic_module_genes[[Hematopoietic_module]], value = sprintf('%s',Hematopoietic_module))

}

Velten_Celltype_Modules <- GSEABase::GeneSetCollection(Hematopoietic_module_genes)

# - - - - without healthies


Velten_Celltype_Modules_scaled_results_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,rownames(Samples_metadata[Samples_metadata$Condition == 'BP-CMML',])]), 
               gene.sets = Velten_Celltype_Modules, cores = 12, ssGSEA.norm = T,
               min.size = 5)

write.csv(Velten_Celltype_Modules_scaled_results_no_healthies, paste(data_folder, '/RNASeq/Pathway_Analysis/ssEnrichments/CellType/2017__Velten_et_al__Healthy_signatures__Enrichments.csv', 
                                    sep = ''),
          quote = F, row.names = T)
```


## 5. Drug Resistance: AZA

```{r}
AZA_response <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Reference_Signatures/Unnikrishnan Ashwin 2017 transcriptome in MDS pre AZA treatment.csv')

AZA_response_UP <- AZA_response %>% dplyr::filter(Comment == 'UP in AZA responders')
AZA_response_UP <- GSEABase::GeneSet(AZA_response_UP$Gene.Symbol..)
AZA_response_UP <- GSEABase::`setName<-`(object = AZA_response_UP, value='AZA response')

AZA_response_DOWN <- AZA_response %>% dplyr::filter(Comment == 'UP in AZA non-responders')
AZA_response_DOWN <- GSEABase::GeneSet(AZA_response_DOWN$Gene.Symbol..)
AZA_response_DOWN <- GSEABase::`setName<-`(object = AZA_response_DOWN, value='AZA resistance')

AZA_response_collection <- GSEABase::GeneSetCollection(AZA_response_UP, AZA_response_DOWN)

# - - - - with healthies

AZA_scaled_result <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = AZA_response_collection, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(AZA_scaled_result, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/AZA_scaled_result_resistance.csv', 
          quote = F, row.names = T)

AZA_result <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = AZA_response_collection, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(AZA_result, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/AZA_result_resistance.csv', 
          quote = F, row.names = T)

# - - - - without healthies

AZA_scaled_result_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = AZA_response_collection, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(AZA_scaled_result_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/AZA_scaled_result_resistance_no_healthies.csv', 
          quote = F, row.names = T)

AZA_result_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = AZA_response_collection, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(AZA_result_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/AZA_result_resistance_no_healthies.csv', 
          quote = F, row.names = T)
```

## 6. Drug Resistance: Venetoclax

```{r}
Veneto_resistance <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Reference_Signatures/Zhang Haijiao 2020 venetoclax resistance signature.csv')

Veneto_resistance <- Veneto_resistance %>% filter(fold_change > 0)

Veneto_resistance$Rank <- 1-percent_rank(Veneto_resistance$fdr)

Veneto_resistance <- Veneto_resistance %>% arrange(fdr)

Veneto_resistance_top_10percent_genes <- Veneto_resistance[grep(TRUE,percent_rank(Veneto_resistance$Rank) >= 0.90),'symbol']
Veneto_resistance_genes <- GSEABase::GeneSet(Veneto_resistance_top_10percent_genes)
Veneto_resistance_genes <- GSEABase::`setName<-`(object = Veneto_resistance_genes, value='Venetoclax resistance')
Veneto_resistance_genes <- GSEABase::GeneSetCollection(Veneto_resistance_genes)

# - - - - with healthies

Veneto_scaled_result <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = Veneto_resistance_genes, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(Veneto_scaled_result, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/Veneto_scaled_result_resistance.csv', 
          quote = F, row.names = T)


Veneto_result <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = Veneto_resistance_genes, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(Veneto_result, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/Veneto_result_resistance.csv', 
          quote = F, row.names = T)


# - - - - without healthies

Veneto_scaled_result_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = Veneto_resistance_genes, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(Veneto_scaled_result_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/Veneto_scaled_result_resistance_no_healthies.csv', 
          quote = F, row.names = T)

Veneto_result_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = Veneto_resistance_genes, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(Veneto_result_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/Veneto_result_resistance_no_healthies.csv', 
          quote = F, row.names = T)
```

## 7. Drug Response: Cytarabine (Ara-C)

```{r}
cytarabine_response <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Reference_Signatures/Xu Haiyan 2019 Cytarabine response signature.csv')

cytarabine_response <- cytarabine_response %>% filter(Log2_FoldChange > 0)

cytarabine_response_genes <- GSEABase::GeneSet(cytarabine_response$GeneName)
cytarabine_response_genes <- GSEABase::`setName<-`(object = cytarabine_response_genes, value='Cytarabine response')
cytarabine_response_genes <- GSEABase::GeneSetCollection(cytarabine_response_genes)

# - - - - with healthies

Cytarabine_scaled_result <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = cytarabine_response_genes, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(Cytarabine_scaled_result, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Response/Cytarabine_scaled_result_response.csv', 
          quote = F, row.names = T)

Cytarabine_result <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = cytarabine_response_genes, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(Cytarabine_result, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Response/Cytarabine_result_response.csv', 
          quote = F, row.names = T)

# - - - - without healthies

Cytarabine_scaled_result_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = cytarabine_response_genes, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(Cytarabine_scaled_result_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Response/Cytarabine_scaled_result_response_no_healthies.csv', 
          quote = F, row.names = T)

Cytarabine_result_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = cytarabine_response_genes, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(Cytarabine_result_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Response/Cytarabine_result_response_no_healthies.csv', 
          quote = F, row.names = T)
```

## 8. Drug Resistance: Daunorubicin

```{r}
Daunorubicin_resistance <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Reference_Signatures/Williams Mark 2020 Daunorubicin resistance signature.csv')

Daunorubicin_resistance$Log2FC <- log2(Daunorubicin_resistance$FC)

Daunorubicin_resistance <- Daunorubicin_resistance %>% filter(Log2FC > 0)

Daunorubicin_resistance_genes <- GSEABase::GeneSet(Daunorubicin_resistance$symbol[which((Daunorubicin_resistance$symbol == '') == F)])

Daunorubicin_resistance_genes <- GSEABase::`setName<-`(object = Daunorubicin_resistance_genes, value='Daunorubicin resistance')

Daunorubicin_resistance_genes <- GSEABase::GeneSetCollection(Daunorubicin_resistance_genes)

# - - - - with healthies

DAC_scaled_result <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = Daunorubicin_resistance_genes, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(DAC_scaled_result, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/DAC_scaled_result_resistance.csv', 
          quote = F, row.names = T)

DAC_result <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = Daunorubicin_resistance_genes, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(DAC_result, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/DAC_result_resistance.csv', 
          quote = F, row.names = T)

# - - - - without healthies

DAC_scaled_result_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = Daunorubicin_resistance_genes, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(DAC_scaled_result_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/DAC_scaled_result_resistance_no_healthies.csv', 
          quote = F, row.names = T)

DAC_result_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = Daunorubicin_resistance_genes, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(DAC_result_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/DAC_result_resistance_no_healthies.csv', 
          quote = F, row.names = T)
```

## 9. Drug Resistance: BEAT AML mined

```{r}
Finaldrugs_selection <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/00__Publicly_Available_Datasets/Vizome__BeatAML_2022/00__Sample_selection_for_Drug_Response_Signatures/Results/Drugs_passing_threshold.csv')$Drug

DrugResistances <- list()
DrugResistances_genes <- list()

for (drug in Finaldrugs_selection) {
  
DrugResistances[[drug]] <- read.csv(paste(sprintf("/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/00__Publicly_Available_Datasets/Vizome__BeatAML_2022/01__DESeq2_Processing/Results/DEGs_ExVivo_Inhibitor_Responsive_vs_Resistant/%s/",drug),sprintf("%s_Responsive_vs_Resistant_upregulated_padj_significant_DEGs_LFC_Ashr.csv", drug), sep = ""))

DrugResistances[[drug]]$rank <- NULL

if ((length(DrugResistances[[drug]]$symbol) < 10) == TRUE) {
DrugResistances[[drug]] <- NULL
} else {
  if ((length(DrugResistances[[drug]]$symbol) > 500) == TRUE) {
      DrugResistances[[drug]]$Rank <- 1-percent_rank(DrugResistances[[drug]]$padj)
      DrugResistances_genes[[drug]] <- DrugResistances[[drug]][grep(TRUE,percent_rank(DrugResistances[[drug]]$Rank) >= 0.70),'symbol']
        if ((length(DrugResistances_genes[[drug]]) > 500) == TRUE) {
          DrugResistances_genes[[drug]] <- GSEABase::GeneSet(na.omit(unique(DrugResistances[[drug]]$symbol[1:500])))
          DrugResistances_genes[[drug]] <- GSEABase::`setName<-`(object = DrugResistances_genes[[drug]], value=sprintf('%s',drug)) 
        }
        else {
          DrugResistances_genes[[drug]] <- GSEABase::GeneSet(na.omit(unique(DrugResistances_genes[[drug]])))
          DrugResistances_genes[[drug]] <- GSEABase::`setName<-`(object = DrugResistances_genes[[drug]], value=sprintf('%s',drug)) 
        }
      }
  else {
      DrugResistances_genes[[drug]] <- GSEABase::GeneSet(na.omit(unique(DrugResistances[[drug]]$symbol[1:500])))
      DrugResistances_genes[[drug]] <- GSEABase::`setName<-`(object = DrugResistances_genes[[drug]], value=sprintf('%s',drug)) 
    }
  }
}

df <- data.frame(row.names = names(DrugResistances_genes))
for (drug in names(DrugResistances_genes)) {
df[drug,'length'] <- length(GSEABase::geneIds(DrugResistances_genes[[drug]]))
}

DrugResistanceSignatures <- GSEABase::GeneSetCollection(DrugResistances_genes)

# - - - - with healthies

DrugResistanceSignatures_scaled_results <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = DrugResistanceSignatures, cores = 6, ssGSEA.norm = T,
               min.size = 10)

colnames(DrugResistanceSignatures_scaled_results) <- gsub("\\.{2}", " ", colnames(DrugResistanceSignatures_scaled_results))
colnames(DrugResistanceSignatures_scaled_results) <- gsub("\\.*$","", colnames(DrugResistanceSignatures_scaled_results))
DrugResistanceSignatures_scaled_results <- DrugResistanceSignatures_scaled_results[,sort(colnames(DrugResistanceSignatures_scaled_results))]

write.csv(DrugResistanceSignatures_scaled_results, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/BeatAMLDrugResistance_scaled_results.csv', 
          quote = F, row.names = T)


DrugResistanceSignatures_results <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = DrugResistanceSignatures, cores = 6, ssGSEA.norm = F,
               min.size = 10)

colnames(DrugResistanceSignatures_results) <- gsub("\\.{2}", " ", colnames(DrugResistanceSignatures_results))
colnames(DrugResistanceSignatures_results) <- gsub("\\.*$","", colnames(DrugResistanceSignatures_results))
DrugResistanceSignatures_results <- DrugResistanceSignatures_results[,sort(colnames(DrugResistanceSignatures_results))]

write.csv(DrugResistanceSignatures_results, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/BeatAMLDrugResistance_results.csv', 
          quote = F, row.names = T)

# - - - - without healthies

DrugResistanceSignatures_scaled_results_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = DrugResistanceSignatures, cores = 6, ssGSEA.norm = T,
               min.size = 10)

colnames(DrugResistanceSignatures_scaled_results_no_healthies) <- gsub("\\.{2}", " ", colnames(DrugResistanceSignatures_scaled_results_no_healthies))
colnames(DrugResistanceSignatures_scaled_results_no_healthies) <- gsub("\\.*$","", colnames(DrugResistanceSignatures_scaled_results_no_healthies))
DrugResistanceSignatures_scaled_results_no_healthies <- DrugResistanceSignatures_scaled_results_no_healthies[,sort(colnames(DrugResistanceSignatures_scaled_results_no_healthies))]

write.csv(DrugResistanceSignatures_scaled_results_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/BeatAMLDrugResistance_scaled_results_no_healthies.csv', 
          quote = F, row.names = T)


DrugResistanceSignatures_results_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = DrugResistanceSignatures, cores = 6, ssGSEA.norm = F,
               min.size = 10)

colnames(DrugResistanceSignatures_results_no_healthies) <- gsub("\\.{2}", " ", colnames(DrugResistanceSignatures_results_no_healthies))
colnames(DrugResistanceSignatures_results_no_healthies) <- gsub("\\.*$","", colnames(DrugResistanceSignatures_results_no_healthies))
DrugResistanceSignatures_results_no_healthies <- DrugResistanceSignatures_results_no_healthies[,sort(colnames(DrugResistanceSignatures_results_no_healthies))]

write.csv(DrugResistanceSignatures_results_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Resistance/BeatAMLDrugResistance_results_no_healthies.csv', 
          quote = F, row.names = T)

```

## 10. Drug Response: BEAT AML mined

```{r}
Finaldrugs_selection <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/00__Publicly_Available_Datasets/Vizome__BeatAML_2022/00__Sample_selection_for_Drug_Response_Signatures/Results/Drugs_passing_threshold.csv')$Drug

DrugResponses <- list()
DrugResponses_genes <- list()

for (drug in Finaldrugs_selection) {
  
DrugResponses[[drug]] <- read.csv(paste(sprintf("/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/00__Publicly_Available_Datasets/Vizome__BeatAML_2022/01__DESeq2_Processing/Results/DEGs_ExVivo_Inhibitor_Responsive_vs_Resistant/%s/",drug),sprintf("%s_Responsive_vs_Resistant_downregulated_padj_significant_DEGs_LFC_Ashr.csv", drug), sep = ""))

DrugResponses[[drug]]$rank <- NULL

if ((length(DrugResponses[[drug]]$symbol) < 10) == TRUE) {
DrugResponses[[drug]] <- NULL 
} else {
  if ((length(DrugResponses[[drug]]$symbol) > 500) == TRUE) {
      DrugResponses[[drug]]$Rank <- 1-percent_rank(DrugResponses[[drug]]$padj)
      DrugResponses_genes[[drug]] <- DrugResponses[[drug]][grep(TRUE,percent_rank(DrugResponses[[drug]]$Rank) >= 0.70),'symbol']
        if ((length(DrugResponses_genes[[drug]]) > 500) == TRUE) {
          DrugResponses_genes[[drug]] <- GSEABase::GeneSet(na.omit(unique(DrugResponses[[drug]]$symbol[1:500])))
          DrugResponses_genes[[drug]] <- GSEABase::`setName<-`(object = DrugResponses_genes[[drug]], value=sprintf('%s',drug)) 
        }
        else {
          DrugResponses_genes[[drug]] <- GSEABase::GeneSet(na.omit(unique(DrugResponses_genes[[drug]])))
          DrugResponses_genes[[drug]] <- GSEABase::`setName<-`(object = DrugResponses_genes[[drug]], value=sprintf('%s',drug)) 
        }
      }
  else {
      DrugResponses_genes[[drug]] <- GSEABase::GeneSet(na.omit(unique(DrugResponses[[drug]]$symbol[1:500])))
      DrugResponses_genes[[drug]] <- GSEABase::`setName<-`(object = DrugResponses_genes[[drug]], value=sprintf('%s',drug)) 
    }
  }
}

df <- data.frame(row.names = names(DrugResponses_genes))
for (drug in names(DrugResponses_genes)) {
df[drug,'length'] <- length(GSEABase::geneIds(DrugResponses_genes[[drug]]))
}

DrugResponseSignatures <- GSEABase::GeneSetCollection(DrugResponses_genes)

# - - - - with healthies

DrugResponseSignatures_scaled_results <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = DrugResponseSignatures, cores = 6, ssGSEA.norm = T,
               min.size = 10)

colnames(DrugResponseSignatures_scaled_results) <- gsub("\\.{2}", " ", colnames(DrugResponseSignatures_scaled_results))
colnames(DrugResponseSignatures_scaled_results) <- gsub("\\.*$","", colnames(DrugResponseSignatures_scaled_results))
DrugResponseSignatures_scaled_results <- DrugResponseSignatures_scaled_results[,sort(colnames(DrugResponseSignatures_scaled_results))]

write.csv(DrugResponseSignatures_scaled_results, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Response/BeatAMLDrugResponseSignatures_Scaled_results.csv', 
          quote = F, row.names = T)



DrugResponseSignatures_results <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = DrugResponseSignatures, cores = 6, ssGSEA.norm = F,
               min.size = 10)

colnames(DrugResponseSignatures_results) <- gsub("\\.{2}", " ", colnames(DrugResponseSignatures_results))
colnames(DrugResponseSignatures_results) <- gsub("\\.*$","", colnames(DrugResponseSignatures_results))
DrugResponseSignatures_results <- DrugResponseSignatures_results[,sort(colnames(DrugResponseSignatures_results))]

write.csv(DrugResponseSignatures_results, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Response/BeatAMLDrugResponseSignatures_results.csv', 
          quote = F, row.names = T)

# - - - - without healthies

DrugResponseSignatures_scaled_results_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = DrugResponseSignatures, cores = 6, ssGSEA.norm = T,
               min.size = 10)

colnames(DrugResponseSignatures_scaled_results_no_healthies) <- gsub("\\.{2}", " ", colnames(DrugResponseSignatures_scaled_results_no_healthies))
colnames(DrugResponseSignatures_scaled_results_no_healthies) <- gsub("\\.*$","", colnames(DrugResponseSignatures_scaled_results_no_healthies))
DrugResponseSignatures_scaled_results_no_healthies <- DrugResponseSignatures_scaled_results_no_healthies[,sort(colnames(DrugResponseSignatures_scaled_results_no_healthies))]

write.csv(DrugResponseSignatures_scaled_results_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Response/BeatAMLDrugResponseSignatures_scaled_results_no_healthies.csv', 
          quote = F, row.names = T)



DrugResponseSignatures_results_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = DrugResponseSignatures, cores = 6, ssGSEA.norm = F,
               min.size = 10)

colnames(DrugResponseSignatures_results_no_healthies) <- gsub("\\.{2}", " ", colnames(DrugResponseSignatures_results_no_healthies))
colnames(DrugResponseSignatures_results_no_healthies) <- gsub("\\.*$","", colnames(DrugResponseSignatures_results_no_healthies))
DrugResponseSignatures_results_no_healthies <- DrugResponseSignatures_results_no_healthies[,sort(colnames(DrugResponseSignatures_results_no_healthies))]

write.csv(DrugResponseSignatures_results_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Drug_Response/BeatAMLDrugResponseSignatures_results_no_healthies.csv', 
          quote = F, row.names = T)
```

## 11. WGCNA score

```{r}
WGCNA_top_genes <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__GRN_Analyses/WGCNA/Results/Representative_Genes/WGCNA_Modules_Top10_percent_genes.csv')

Top_Brown <- WGCNA_top_genes %>% dplyr::filter(Module == 'Brown')
Top_Brown <- GSEABase::GeneSet(Top_Brown$Gene)
Top_Brown <- GSEABase::`setName<-`(object = Top_Brown, value='Brown Module')

Top_Blue <- WGCNA_top_genes %>% dplyr::filter(Module == 'Blue')
Top_Blue <- GSEABase::GeneSet(unique(Top_Blue$Gene))
Top_Blue <- GSEABase::`setName<-`(object = Top_Blue, value='Blue Module')

Top_Red <- WGCNA_top_genes %>% dplyr::filter(Module == 'Red')
Top_Red <- GSEABase::GeneSet(Top_Red$Gene)
Top_Red <- GSEABase::`setName<-`(object = Top_Red, value='Red Module')

Top_Black <- WGCNA_top_genes %>% dplyr::filter(Module == 'Black')
Top_Black <- GSEABase::GeneSet(Top_Black$Gene)
Top_Black <- GSEABase::`setName<-`(object = Top_Black, value='Black Module')

Top_Yellow <- WGCNA_top_genes %>% dplyr::filter(Module == 'Yellow')
Top_Yellow <- GSEABase::GeneSet(Top_Yellow$Gene)
Top_Yellow <- GSEABase::`setName<-`(object = Top_Yellow, value='Yellow Module')

Top_Turquoise <- WGCNA_top_genes %>% dplyr::filter(Module == 'Turquoise')
Top_Turquoise <- GSEABase::GeneSet(Top_Turquoise$Gene)
Top_Turquoise <- GSEABase::`setName<-`(object = Top_Turquoise, value='Turquoise Module')

Top_Green <- WGCNA_top_genes %>% dplyr::filter(Module == 'Green')
Top_Green <- GSEABase::GeneSet(Top_Green$Gene)
Top_Green <- GSEABase::`setName<-`(object = Top_Green, value='Green Module')

Top_Grey <- WGCNA_top_genes %>% dplyr::filter(Module == 'Grey')
Top_Grey <- GSEABase::GeneSet(Top_Grey$Gene)
Top_Grey <- GSEABase::`setName<-`(object = Top_Grey, value='Grey Module')

WGCNA_modules_top_genes <- GSEABase::GeneSetCollection(Top_Grey, 
                                                       Top_Green,
                                                       Top_Turquoise,
                                                       Top_Blue,
                                                       Top_Red,
                                                       Top_Brown,
                                                       Top_Black,
                                                       Top_Yellow)

# - - - - with healthies

WGCNA_scaled_score <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = WGCNA_modules_top_genes, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(WGCNA_scaled_score, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/WGCNA/WGCNA_Module_Scaled_Scores.csv', 
          quote = F, row.names = T)

WGCNA_score <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = WGCNA_modules_top_genes, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(WGCNA_score, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/WGCNA/WGCNA_Module_Scores.csv', 
          quote = F, row.names = T)

# - - - - without healthies

WGCNA_scaled_score_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = WGCNA_modules_top_genes, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(WGCNA_scaled_score_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/WGCNA/WGCNA_Module_Scaled_Scores_No_Healthies.csv', 
          quote = F, row.names = T)


WGCNA_score_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = WGCNA_modules_top_genes, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(WGCNA_score_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/WGCNA/WGCNA_Module_Scores_No_Healthies.csv', 
          quote = F, row.names = T)

######
# Plotting WGCNA Score distributions
######

library(ggplot2)
library(ggridges)
library(dplyr)

WGCNA_scaled_score <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/WGCNA/WGCNA_Module_Scaled_Scores.csv')

df <- data.frame()

df[1:50,'Module'] <- rep('Grey.Module', n = 50)
df[1:50,'Sample'] <- WGCNA_scaled_score$X
df[1:50,'ssGSEA Score'] <- WGCNA_scaled_score$Grey.Module

df[51:100,'Module'] <- rep('Green.Module', n = 50)
df[51:100,'Sample'] <- WGCNA_scaled_score$X
df[51:100,'ssGSEA Score'] <- WGCNA_scaled_score$Green.Module

df[101:150,'Module'] <- rep('Turquoise.Module', n = 50)
df[101:150,'Sample'] <- WGCNA_scaled_score$X
df[101:150,'ssGSEA Score'] <- WGCNA_scaled_score$Turquoise.Module

df[151:200,'Module'] <- rep('Blue.Module', n = 50)
df[151:200,'Sample'] <- WGCNA_scaled_score$X
df[151:200,'ssGSEA Score'] <- WGCNA_scaled_score$Blue.Module

df[201:250,'Module'] <- rep('Red.Module', n = 50)
df[201:250,'Sample'] <- WGCNA_scaled_score$X
df[201:250,'ssGSEA Score'] <- WGCNA_scaled_score$Red.Module

df[251:300,'Module'] <- rep('Brown.Module', n = 50)
df[251:300,'Sample'] <- WGCNA_scaled_score$X
df[251:300,'ssGSEA Score'] <- WGCNA_scaled_score$Brown.Module

df[301:350,'Module'] <- rep('Black.Module', n = 50)
df[301:350,'Sample'] <- WGCNA_scaled_score$X
df[301:350,'ssGSEA Score'] <- WGCNA_scaled_score$Black.Module

df[351:400,'Module'] <- rep('Yellow.Module', n = 50)
df[351:400,'Sample'] <- WGCNA_scaled_score$X
df[351:400,'ssGSEA Score'] <- WGCNA_scaled_score$Yellow.Module

df[grep(TRUE,!df$Sample %in% c('929','1126','1138','1157','1162','1170','1207')),'Condition'] <- 
'BP-CMML'
df[grep(FALSE,!df$Sample %in% c('929','1126','1138','1157','1162','1170','1207')),'Condition'] <- 
'Control'


png(filename = '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/WGCNA/WGCNA_distributions.png', units = 'cm', height = 15.77, width = 7.61, res = 300)

 ggplot(df, 
        aes(x = `ssGSEA Score`, y = Module, 
            color = Condition, point_color = Condition, fill = Condition)) +
  geom_density_ridges(
    jittered_points = TRUE, scale = .95, rel_min_height = .01,
    point_shape = "|", point_size = 3, size = 0.5,
    position = position_points_jitter(height = 0)
    ) +
  scale_fill_manual(values = c(alpha("#DD3429",0.75), alpha("#A4BECE",0.75)), 
                    labels = c("BP-CMML", "Control"), ) +
  scale_color_manual(values = c("#DD3429", "#A4BECE"), guide = "none") +
  scale_discrete_manual("point_color", values = c("#DD3429", "#A4BECE"), guide = "none") +
  coord_cartesian(clip = "off") +
  guides(fill = guide_legend(
    override.aes = list(
      fill = c("#DD3429", "#A4BECE"),
      color = NA, point_color = NA)
    )
  ) +
  ggtitle("") + ylab("") +
  theme_ridges(center = TRUE) + theme_minimal() + 
   theme(axis.title.x = element_text(face='bold', size=10),
        legend.position = 'bottom',
        legend.title = element_text(face='bold', size=12)) + guides(fill = 
                                          guide_legend(title = 'Condition',
                                  title.position = "top",title.hjust = 0.5))
 
 dev.off()
```

## 12. Mutation associated genes

```{r}
Mutation_associated_genes <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Reference_Signatures/Mutation_associated_pathways.csv')

SRSF2_sig <- GSEABase::GeneSet(na.omit(unique(Mutation_associated_genes$SRSF2.group)))
SRSF2_sig <- GSEABase::`setName<-`(object = SRSF2_sig, value='SRSF2 signature')

RUNX1_sig <- GSEABase::GeneSet(na.omit(unique(Mutation_associated_genes$RUNX1.group)))
RUNX1_sig <- GSEABase::`setName<-`(object = RUNX1_sig, value='RUNX1 signature')

RAS_sig <- GSEABase::GeneSet(na.omit(unique(Mutation_associated_genes$RAS.group)))
RAS_sig <- GSEABase::`setName<-`(object = RAS_sig, value='RAS signature')

FLT3_ITD_sig <- GSEABase::GeneSet(na.omit(unique(Mutation_associated_genes$ITD.group)))
FLT3_ITD_sig <- GSEABase::`setName<-`(object = FLT3_ITD_sig, value='FLT3 ITD signature')

NPM1_sig <- GSEABase::GeneSet(na.omit(unique(Mutation_associated_genes$NPM1.group)))
NPM1_sig <- GSEABase::`setName<-`(object = NPM1_sig, value='NPM1 signature')

FLT3_ITD_NPM1_sig <- GSEABase::GeneSet(na.omit(unique(Mutation_associated_genes$ITD.NPM1.group)))
FLT3_ITD_NPM1_sig <- GSEABase::`setName<-`(object = FLT3_ITD_NPM1_sig, value='FLT3 ITD NPM1 signature')

Mutation_signatures <- GSEABase::GeneSetCollection(SRSF2_sig, 
                                                    RUNX1_sig,
                                                    RAS_sig,
                                                    FLT3_ITD_sig,
                                                    NPM1_sig,
                                                   FLT3_ITD_NPM1_sig)

# - - - - with healthies

Mutation_signatures_scaled_score <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = Mutation_signatures, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(Mutation_signatures_scaled_score, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Others/Cellcycle_sig_Scaled_Scores.csv', 
          quote = F, row.names = T)

Mutation_signatures_score <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = Mutation_signatures, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(Mutation_signatures_score, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Others/Cellcycle_sig_Scores.csv', 
          quote = F, row.names = T)

# - - - - without healthies

Mutation_signatures_scaled_score_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = Mutation_signatures, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(Mutation_signatures_scaled_score_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Others/Mutation_signatures_Scaled_Scores_no_healthies.csv', 
          quote = F, row.names = T)


Mutation_signatures_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = Mutation_signatures, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(Mutation_signatures_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Others/Mutation_signatures_Scores_no_healthies.csv', 
          quote = F, row.names = T)
```

  ## 13. Other signatures

```{r}
Cellcycle_sig <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Reference_Signatures/Roy 2021 Transitions in lineage signatures.csv')

Quiescence_sig <- GSEABase::GeneSet(unique(Cellcycle_sig$Quiescence))
Quiescence_sig <- GSEABase::`setName<-`(object = Quiescence_sig, value='Quiescence')

G2M_sig <- GSEABase::GeneSet(unique(Cellcycle_sig$G2M))
G2M_sig <- GSEABase::`setName<-`(object = G2M_sig, value='G2M')

S_phase_sig <- GSEABase::GeneSet(unique(Cellcycle_sig$S_Phase))
S_phase_sig <- GSEABase::`setName<-`(object = S_phase_sig, value='S_phase')

Cellcycle_signatures <- GSEABase::GeneSetCollection(Quiescence_sig, 
                                                       G2M_sig,
                                                       S_phase_sig)

# - - - - with healthies

Cellcycle_signatures_scaled_score <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = Cellcycle_signatures, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(Cellcycle_signatures_scaled_score, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Others/Cellcycle_sig_Scaled_Scores.csv', 
          quote = F, row.names = T)

WGCNA_score <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = WGCNA_modules_top_genes, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(WGCNA_score, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/WGCNA/WGCNA_Module_Scores.csv', 
          quote = F, row.names = T)

# - - - - without healthies

Cellcycle_signatures_scaled_score_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = Cellcycle_signatures, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(Cellcycle_signatures_scaled_score_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Others/Cellcycle_sig_Scaled_Scores_no_healthies.csv', 
          quote = F, row.names = T)


WGCNA_score_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = WGCNA_modules_top_genes, cores = 6, ssGSEA.norm = F,
               min.size = 5)

write.csv(WGCNA_score_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/WGCNA/WGCNA_Module_Scores_No_Healthies.csv', 
          quote = F, row.names = T)

```

### - BCL2 family

```{r}
BCL2_family <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Reference_Signatures/BCL2.txt', sep = '\t')

BCL2_family_sig <- GSEABase::GeneSet(unique(BCL2_family$Approved.symbol))
BCL2_family_sig <- GSEABase::`setName<-`(object = BCL2_family_sig, value='BCL2 family')

BCL2_signatures <- GSEABase::GeneSetCollection(BCL2_family_sig)

# - - - - with healthies

BCL2_signatures_scaled_score <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = BCL2_signatures, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(BCL2_signatures_scaled_score, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Others/BCL2_sig_Scaled_Scores.csv', 
          quote = F, row.names = T)

# - - - - without healthies

BCL2_signatures_scaled_score_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = BCL2_signatures, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(BCL2_signatures_scaled_score_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Others/BCL2_sig_Scaled_Scores_no_healthies.csv', 
          quote = F, row.names = T)

```

### - FOX genes

```{r}
FOX_family <- read.csv('/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Reference_Signatures/FOX.txt', sep = '\t')

FOX_family_sig <- GSEABase::GeneSet(unique(FOX_family$Approved.symbol))
FOX_family_sig <- GSEABase::`setName<-`(object = FOX_family_sig, value='FOX family')

FOX_signatures <- GSEABase::GeneSetCollection(FOX_family_sig)

# - - - - with healthies

FOX_signatures_scaled_score <- enrichIt(obj = as.matrix(Expression_matrix), 
               gene.sets = FOX_signatures, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(FOX_signatures_scaled_score, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Others/FOX_sig_Scaled_Scores.csv', 
          quote = F, row.names = T)

# - - - - without healthies

FOX_signatures_scaled_score_no_healthies <- enrichIt(obj = as.matrix(Expression_matrix[,grep(TRUE,!colnames(Expression_matrix) %in% c("929","1126","1138","1157","1162","1170","1207"))]), 
               gene.sets = FOX_signatures, cores = 6, ssGSEA.norm = T,
               min.size = 5)

write.csv(FOX_signatures_scaled_score_no_healthies, '/Users/kristiangurashi/Dropbox (The University of Manchester)/Projects data/Project__Christie_BP_CMML_Stratification_and_Drug_Prediction/04__Christie_BP_CMML_BulkRNAseq/02__Pathway_Analyses/Results/Others/FOX_sig_Scaled_Scores_no_healthies.csv', 
          quote = F, row.names = T)
```
