---
title: "Figure 05 - Panels HI"
output: 
  rmdformats::robobook:
    lightbox: true
    fig_caption: true
---

```{=html}
<style type="text/css">
  body{
  font-size: 10pt;
}
</style>
```

# 1) Setup

## - Defining work directory

In this section, we define the working directory for the R Markdown document.

```{r setup, include=TRUE}
# Get the path of the current script
# Then get the parent directory of the parent directory of the parent directory
local_wd_folder <- dirname(dirname(dirname(rstudioapi::getSourceEditorContext()$path)))

# Set the root directory for knitr to the local working directory
knitr::opts_knit$set(root.dir = local_wd_folder)
```

## - Defining input data and output directories

Here, we define the directories for input data and output files.

```{r}
# Get the directory of the current script
script_folder <- dirname(rstudioapi::getSourceEditorContext()$path)

# Define the data folder and output folder
data_folder <- './Data'
#results_folder <- './Results'
figures_folder <- './Figures'

dir.create(paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/C1_vs_C2', sep = ''), 
           showWarnings = F)
```

## - Setting seed

Setting a seed ensures that any random processes are reproducible.

```{r}
# Set a seed for reproducibility
set.seed(123)
```

## - Packages installation (optional)

In this section, we ensure that all necessary packages are installed.

```{r}
# Ensure BiocManager is available for installation of Bioconductor packages
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(version = "3.18", ask = FALSE)

# Define a list of required packages used in this script
packages_required <- c("ggplot2", "dplyr", "fgsea",
                       "unikn", "RColorBrewer", "yarrr", "scales", "ggsci")

# Identify any required packages that are not installed
packages_uninstalled <- packages_required[!(packages_required %in% installed.packages()[,"Package"])]

# Install any uninstalled packages
if(length(packages_uninstalled)) BiocManager::install(packages_uninstalled, ask = FALSE)
```

## - Loading packages

Here, we load the necessary packages for our analysis.

```{r message=FALSE, warning=FALSE}
library(fgsea, quietly = TRUE)
library(ggplot2, quietly = TRUE)
library(dplyr, quietly = TRUE)
```

## - Loading palettes

In this section, we load additional color palettes and define some custom ones.

```{r message=FALSE, warning=FALSE}
# Load additional colour palette packages
library(unikn, quietly = TRUE)
library(RColorBrewer, quietly = TRUE)
library(yarrr, quietly = TRUE)
library(scales, quietly = TRUE)
library(ggsci, quietly = TRUE)

# Define a set of custom color palettes from the unikn package
mix_1 <- usecol(pal = c(Karpfenblau, "white", Peach), n = 15)
mix_2 <- usecol(pal = c(rev(pal_seeblau), "white", pal_pinky))
mix_3 <- usecol(pal = c(rev(pal_bordeaux), "white", pal_petrol), n = 15)

# Display the custom palettes
seecol(list(mix_1, mix_2, mix_3), col_brd = "white", lwd_brd = 4, title = "Comparing palettes mixed from unikn colors", pal_names = c("mix_1", "mix_2", "mix_3"))

# Define a second set of custom palettes from the RColorBrewer and yarrr packages
brew_mix <- usecol(c(rev(brewer.pal(n = 4, name = "Reds")), "white", brewer.pal(n = 4, name = "Blues")), n = 13)
brew_ext <- usecol(brewer.pal(n = 11, name = "Spectral"), n = 12)
yarrr_mix <- usecol(c(piratepal("nemo"), piratepal("bugs")))
yarrr_mod <- usecol(c(piratepal("ipod")), n = 9)

# Display the second set of custom palettes
seecol(pal = list(brew_mix, brew_ext, yarrr_mix, yarrr_mod), col_brd = "white", lwd_brd = 2, title = "Using usecol() and seecol() to mix and modify palettes", pal_names = c("brew_mix", "brew_ext", "yarrr_mix", "yarrr_mod"))

# Define additional custom palettes from the scales package
natjournals_palette <- pal_npg("nrc")(10)
```

## - Log Session Info

Finally, we log the session information for reproducibility.

```{r message=TRUE, warning=TRUE}
# Write the session information to a text file
writeLines(capture.output(sessionInfo()), file.path(script_folder, 'Fig_05_HI_SessionInfo.txt'))

# Print the session information
sessionInfo()
```

# 2) fGSEA Analyses 

## - Gene ranking

Here we read in a CSV file containing differentially expressed genes (DEGs) between the blast phase and control. We then calculate a rank for each gene based on its log2 fold change and p-value. The DEGs are arranged by their rank. Finally, we prepare a named numeric vector of ranks, removing any NA values.

```{r}
# Read DEGs
C1_vs_C2_DEGS <- read.csv(
  file = paste(data_folder, '/RNASeq/InHouse/DESeq2/DEGs_Comparisons/C1_vs_C2/DGE_C1_vs_C2_ALL_DEGs_LFC_Ashr.csv', sep = ''), 
  row.names = 'X'
)

# Calculate rank
C1_vs_C2_DEGS$Rank <- sign(C1_vs_C2_DEGS$log2FoldChange)*(-1*log10(C1_vs_C2_DEGS$pvalue))

# Arrange by rank
C1_vs_C2_DEGS <- C1_vs_C2_DEGS %>% arrange(-Rank)

# Prepare ranks
Ranks <- C1_vs_C2_DEGS$Rank
names(Ranks) <- C1_vs_C2_DEGS$GeneSymbol
Ranks <- Ranks[!is.na(Ranks)]
```

## - Hallmark analysis

Here we read in a GMT file containing Hallmark pathways. We then perform a fast gene set enrichment analysis (fgsea) using the Hallmark pathways and the ranks of the DEGs. The results are arranged by the adjusted p-value.

We then write the full results to a CSV file. We also write two additional CSV files: one containing the pathways that are upregulated in the Immature cluster (C1) and one containing the pathways that are upregulated in the Mature cluster (C2).

```{r}
# Read Hallmark pathways
Hallmark_pathways_gmt_file <- paste(data_folder, '/Resources/RNASeq/Gene_Pathways/2022__GSEA_MSIGDB__h.all.v2022.1.Hs.symbols.gmt.txt', sep = '')
Hallmark_pathways <- gmtPathways(Hallmark_pathways_gmt_file)
str(head(Hallmark_pathways))

# Perform fgsea
fgseaRes_Hallmark <- fgsea(
  pathways = Hallmark_pathways, 
  stats = Ranks, 
  minSize=10, 
  maxSize = 500, 
  nPermSimple = 10000
)

# Arrange by padj
fgseaRes_Hallmark <- fgseaRes_Hallmark %>% arrange(padj)
fgseaRes_Hallmark$leadingEdge <- NULL


# Write results to csv
write.csv(
  x = fgseaRes_Hallmark,
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/C1_vs_C2/fGSEA_H_C1_vs_C2_all.csv', sep = '')
)

# Write upregulated in BP-CMML to csv
write.csv(
  x = fgseaRes_Hallmark %>% filter(NES > 0),
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/C1_vs_C2/fGSEA_H_C1_vs_C2__Up_in_BP-CMML.csv', sep = '')
)

# Write upregulated in Control to csv
write.csv(
  x = fgseaRes_Hallmark %>% filter(NES < 0),
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/C1_vs_C2/fGSEA_H_C1_vs_C2__Up_in_Control.csv', sep = '')
)
```

## - Canonical pathway analysis

Here we read in a GMT file containing Canonical pathways. We then perform a fast gene set enrichment analysis (fgsea) using the Canonical pathways and the ranks of the DEGs. The results are arranged by the adjusted p-value.

We then write the full results to a CSV file. We also write two additional CSV files: one containing the pathways that are upregulated in the Immature cluster (C1) and one containing the pathways that are upregulated in the Mature cluster (C2).

```{r}
# Read Canonical pathways
CanonicalPathways_gmt_file <- paste(data_folder, '/Resources/RNASeq/Gene_Pathways/2022__GSEA_MSIGDB__c2.cp.v2022.1.Hs.symbols.gmt.txt', sep = '')
CanonicalPathways <- gmtPathways(CanonicalPathways_gmt_file)
str(head(CanonicalPathways))


# Perform fgsea
fgseaRes_CP <- fgsea(
  pathways = CanonicalPathways, 
  stats = Ranks, 
  minSize=10, 
  maxSize = 500, 
  nPermSimple = 10000
)

# Arrange by padj
fgseaRes_CP <- fgseaRes_CP %>% arrange(padj)
fgseaRes_CP$leadingEdge <- NULL

# Write results to csv
write.csv(
  x = fgseaRes_CP,
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/C1_vs_C2/fGSEA_CP_C1_vs_C2_all.csv', sep = '')
)

# Write upregulated in BP-CMML to csv
write.csv(
  x = fgseaRes_CP %>% filter(NES > 0), 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/C1_vs_C2/fGSEA_CP_C1_vs_C2__Up_in_BP-CMML.csv', sep = '')
)

# Write upregulated in Control to csv
write.csv(
  x = fgseaRes_CP %>% filter(NES < 0), 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/C1_vs_C2/fGSEA_CP_C1_vs_C2__Up_in_Control.csv', sep = '')
)
```

## - GO BP pathway analysis

Here we read in a GMT file containing Gene Ontology Biological Process (GO_BP) pathways. We then perform a fast gene set enrichment analysis (fgsea) using the GO_BP pathways and the ranks of the DEGs. The results are arranged by the adjusted p-value.

We then write the full results to a CSV file. We also write two additional CSV files: one containing the pathways that are upregulated in the Immature cluster (C1) and one containing the pathways that are upregulated in the Mature cluster (C2).

```{r}
# Read GO_BP pathways
GO_BP_gmt_file <- paste(data_folder, '/Resources/RNASeq/Gene_Pathways/2022__GSEA_MSIGDB__c5.go.bp.v2022.1.Hs.symbols.gmt.txt', sep = '')
GOBP_pathways <- gmtPathways(GO_BP_gmt_file)
str(head(GOBP_pathways))

# Perform fgsea
fgseaRes_GO_BP <- fgsea(
  pathways = GOBP_pathways, 
  stats = Ranks, 
  minSize=10, 
  maxSize = 500, 
  nPermSimple = 10000
)

# Arrange by padj
fgseaRes_GO_BP <- fgseaRes_GO_BP %>% arrange(padj)
fgseaRes_GO_BP$leadingEdge <- NULL

# Write results to csv
write.csv(
  x = fgseaRes_GO_BP, 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/C1_vs_C2/fGSEA_GO_BP_C1_vs_C2_all.csv', sep = '')
)

# Write upregulated in BP-CMML to csv
write.csv(
  x = fgseaRes_GO_BP %>% filter(NES > 0), 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/C1_vs_C2/fGSEA_GO_BP_C1_vs_C2__Up_in_BP-CMML.csv', sep = '')
)

# Write upregulated in Control to csv
write.csv(
  x = fgseaRes_GO_BP %>% filter(NES < 0), 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/C1_vs_C2/fGSEA_GO_BP_C1_vs_C2__Up_in_Control.csv', sep = '')
)
```

## - Pathway plotting

Here we combine gene set enrichment analysis results from different pathway sets, filter for specific pathways of interest, and create a bar plot of the Normalized Enrichment Score (NES) for each pathway, colored by cluster. The plot is then saved as a PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 5H', fig.pos='H', fig.height=4.0708661, fig.width=4.527559}
# Combine fgsea results into one data frame and sort by padj
Cumulative_fGSEAres <- rbind(fgseaRes_CP, fgseaRes_Hallmark, fgseaRes_GO_BP) %>% arrange(padj)

# Filter for positive NES and specific pathways
Cumulative_fGSEAres_posNES <- Cumulative_fGSEAres %>% 
  filter(NES > 0,
         pathway %in% c('KEGG_RIBOSOME',
                        'WP_CYTOPLASMIC_RIBOSOMAL_PROTEINS',
                        'GOBP_RIBOSOME_BIOGENESIS',
                        'GOBP_NEGATIVE_REGULATION_OF_HEMATOPOIETIC_PROGENITOR_CELL_DIFFERENTIATION',
                        'GOBP_HEMATOPOIETIC_STEM_CELL_PROLIFERATION',
                        'GOBP_POSITIVE_REGULATION_OF_MEGAKARYOCYTE_DIFFERENTIATION',
                        'REACTOME_REGULATION_OF_RUNX1_EXPRESSION_AND_ACTIVITY',
                        'REACTOME_RUNX1_REGULATES_GENES_INVOLVED_IN_MEGAKARYOCYTE_DIFFERENTIATION_AND_PLATELET_FUNCTION'))

# Filter for negative NES and specific pathways
Cumulative_fGSEAres_negNES <- Cumulative_fGSEAres %>% 
  filter(NES < 0,
         pathway %in% c('HALLMARK_IL6_JAK_STAT3_SIGNALING',
                        'GOBP_MONONUCLEAR_CELL_DIFFERENTIATION',
                        'HALLMARK_UNFOLDED_PROTEIN_RESPONSE',
                        'HALLMARK_MTORC1_SIGNALING',
                        'KEGG_LYSOSOME',
                        'HALLMARK_OXIDATIVE_PHOSPHORYLATION',
                        'GOBP_REACTIVE_OXYGEN_SPECIES_METABOLIC_PROCESS',
                        'HALLMARK_FATTY_ACID_METABOLISM'))

# Combine positive and negative NES results
Cumulative_fGSEAres_top10 <- rbind(Cumulative_fGSEAres_posNES, Cumulative_fGSEAres_negNES)

# Assign cluster labels
Cumulative_fGSEAres_top10[1:8,'Cluster'] <- 'C1: Immature'
Cumulative_fGSEAres_top10[9:16,'Cluster'] <- 'C2: Mature'

# Rename pathways for readability
Cumulative_fGSEAres_top10$pathway <-  c('Ribosome',
                                       'Cytoplamic ribosomal proteins',
                                       'Ribosome biogenesis',
                                       'Negative regulation HSC diff.',
                                       'HSC proliferation',
                                       "Positive megakaryocyte diff.",
                                       'RUNX1 expression',
                                       'RUNX1 mediated megakaryocyte diff.',
                                       'IL6 JAK2 STAT3',
                                       'MNC differentiation',
                                       'Unfolded protein response (UPR)',
                                       'MTORC1 signaling',
                                       'Lysosome',
                                       'OXPHOS signaling',
                                       'ROS metabolism',
                                       'Fatty acids metabolism')

# Create plot
plot <- ggplot(Cumulative_fGSEAres_top10, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=Cluster)) +
  coord_flip() +
  labs(x="", y="Normalized Enrichment\nScore (NES)",
       title="") + 
  theme_minimal(base_line_size = 0.3) + 
  scale_fill_manual(values = c('C1: Immature'    = '#793F0B',
                               'C2: Mature'    = '#0B544B')) +
  theme(legend.position = 'none',
        plot.title = element_text(face='bold', size =14, family = 'sans'),
        axis.title.y = element_text(size =12, family = 'sans'),
        axis.title.x = element_text(size =12, family = 'sans'))

# Print plot
print(plot)

# Save plot as PNG
png(filename = paste(figures_folder, '/Fig_05/Panel_H.png', sep = ''),
    res = 300, units = 'cm', height = 10.34, width  = 11.5)
print(plot)
invisible(dev.off())
```

## - Enrichment plots selected signatures

Here we read in a GMT file containing pathways related to quiescent vs dividing hematopoietic stem cells (HSCs). We then perform a fast gene set enrichment analysis (fgsea) using these pathways and the ranks of the DEGs.

We then plot the enrichment of the pathway with the smallest p-value and save the plot as a PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 5I', fig.pos='H', fig.height=2.625984, fig.width=3.46457}
# Define file path for Quiescence_HSC gmt file
Quiescence_HSC_gmt_file <- paste(data_folder, 
                        '/Resources/RNASeq/Gene_Pathways/2007__Graham_et_al__NormalQuiescent_vs_NormalDividing.gmt', 
                        sep = '')

# Read the gmt file into a list of pathways
Quiescence_HSC_pathways <- gmtPathways(Quiescence_HSC_gmt_file)

# Display the structure of the first few pathways
str(head(Quiescence_HSC_pathways))

# Perform fgsea on the pathways
fgseaRes_Quiescence_HSC_pathways <- fgsea(pathways = Quiescence_HSC_pathways, 
                           stats = Ranks, 
                           minSize=10, 
                           maxSize = 500, 
                           nPermSimple = 10000)

# Display the fgsea results
fgseaRes_Quiescence_HSC_pathways

# Plot the enrichment of the pathway with the smallest p-value
plot <- plotEnrichment(Quiescence_HSC_pathways[[head(fgseaRes_Quiescence_HSC_pathways[order(pval), ], 1)$pathway]],
               Ranks) + labs(title=head(fgseaRes_Quiescence_HSC_pathways[order(pval), ], 1)$pathway) + 
  ggtitle('UP in quiescence vs dividing HSC\nGraham et al. 2007') + 
  theme(plot.title = element_text(hjust = 0.5, face = 'bold', size = 13, family='sans')) +
  ylab('Enrichment score') + xlab('Rank')

# Print the plot
print(plot)

# Save the plot as a PNG file
png(filename = paste(figures_folder, '/Fig_05/Panel_I-1.png', sep = ''),
    res = 300, units = 'cm', height = 6.67, width  = 8.8)
print(plot)
invisible(dev.off())
```

Here we read in a CSV file containing cell cycle-related gene signatures, perform a fast gene set enrichment analysis (fgsea) on these signatures. We also plot the enrichment of the G2M Phase signature in the results, saving the plot as PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 5I', fig.pos='H', fig.height=2.625984, fig.width=3.46457}
# Define file path for Roy2021 signatures csv file
Roy2021_signatures <- read.csv(paste(data_folder, 
                        '/Resources/RNASeq/Gene_Pathways/2021__Roy_et_al__cellcycle.csv', 
                        sep = ''))

# Convert the dataframe to a list and remove duplicate entries
Roy2021_signatures <- as.list(Roy2021_signatures)
Roy2021_signatures$G2M <- unique(Roy2021_signatures$G2M)[1:length(unique(Roy2021_signatures$G2M))-1]
Roy2021_signatures$S_Phase <- unique(Roy2021_signatures$S_Phase)[1:length(unique(Roy2021_signatures$S_Phase))-1]

# Perform fgsea on the signatures
Roy2021_signatures_results <- fgsea(pathways = Roy2021_signatures, 
                           stats = Ranks, 
                           minSize=10, 
                           maxSize = 500, 
                           nPermSimple = 10000)

# Display the fgsea results
Roy2021_signatures_results

# Plot the enrichment of the signature with the smallest p-value
plot <- plotEnrichment(Roy2021_signatures[[head(Roy2021_signatures_results[order(pval), ], 1)$pathway]],
               Ranks) + labs(title=head(Roy2021_signatures_results[order(pval), ], 1)$pathway) + 
  ggtitle('G2M phase\nTirosh et al. 2016') + 
  theme(plot.title = element_text(hjust = 0.5, face = 'bold', size = 13, family='sans')) +
  ylab('Enrichment score') + xlab('Rank')

# Print the plot
print(plot)

# Save the plot as a PNG file
png(filename = paste(figures_folder, '/Fig_05/Panel_I-2.png', sep = ''),
    res = 300, units = 'cm', height = 6.67, width  = 8.8)
print(plot)
invisible(dev.off())
```

Here we create a plot of the enrichment of the S Phase signature in the cell cycle-related gene signatures results. We then save this plot as a PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 5I', fig.pos='H', fig.height=2.625984, fig.width=3.46457}
# Plot the enrichment of the third pathway in the Roy2021_signatures_results
plot <- plotEnrichment(Roy2021_signatures[[Roy2021_signatures_results[3,'pathway'][[1]]]],
               Ranks) + labs(title='') + 
  ggtitle('S phase\nTirosh et al. 2016') + 
  theme(plot.title = element_text(hjust = 0.5, face = 'bold', size = 13, family='sans')) +
  ylab('Enrichment score') + xlab('Rank')

# Print the plot
print(plot)

# Save the plot as a PNG file
png(filename = paste(figures_folder, '/Fig_05/Panel_I-3.png', sep = ''),
    res = 300, units = 'cm', height = 6.67, width  = 8.8)

print(plot)

invisible(dev.off())
```

Here we create a plot of the enrichment of the quiescence signature in the cell cycle-related gene signatures results. We then save this plot as a PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 5I', fig.pos='H', fig.height=2.625984, fig.width=3.46457}
# Plot the enrichment of the second pathway in the Roy2021_signatures_results
plot <- plotEnrichment(Roy2021_signatures[[Roy2021_signatures_results[2,'pathway'][[1]]]],
               Ranks) + labs(title='') + 
  ggtitle('Quiescence\nGiustacchini et al. 2017') + 
  theme(plot.title = element_text(hjust = 0.5, face = 'bold', size = 13, family='sans')) +
  ylab('Enrichment score') + xlab('Rank')

# Print the plot
print(plot)

# Save the plot as a PNG file
png(filename = paste(figures_folder, '/Fig_05/Panel_I-4.png', sep = ''),
    res = 300, units = 'cm', height = 6.67, width  = 8.8)

print(plot)

invisible(dev.off())
```
