---
title: "Figure 02 - Panels GH + Fig S5"
output: 
  rmdformats::robobook:
    lightbox: true
    fig_caption: true
---

```{=html}
<style type="text/css">
  body{
  font-size: 10pt;
}
</style>
```

# 1) Setup

## - Defining work directory

In this section, we define the working directory for the R Markdown document.

```{r setup, include=TRUE}
# Get the path of the current script
# Then get the parent directory of the parent directory of the parent directory
local_wd_folder <- dirname(dirname(dirname(rstudioapi::getSourceEditorContext()$path)))

# Set the root directory for knitr to the local working directory
knitr::opts_knit$set(root.dir = local_wd_folder)
```

## - Defining input data and output directories

Here, we define the directories for input data and output files.

```{r}
# Get the directory of the current script
script_folder <- dirname(rstudioapi::getSourceEditorContext()$path)

# Define the data folder and output folder
data_folder <- './Data'
#results_folder <- './Results'
figures_folder <- './Figures'
```

## - Setting seed

Setting a seed ensures that any random processes are reproducible.

```{r}
# Set a seed for reproducibility
set.seed(123)
```

## - Packages installation (optional)

In this section, we ensure that all necessary packages are installed.

```{r}
# Ensure BiocManager is available for installation of Bioconductor packages
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(version = "3.18", ask = FALSE)

# Define a list of required packages used in this script
packages_required <- c("ggplot2", "dplyr", "fgsea", "escape", "hrbrthemes", 
                       "tidyr", "ggpubr",
                       "unikn", "RColorBrewer", "yarrr", "scales", "ggsci")

# Identify any required packages that are not installed
packages_uninstalled <- packages_required[!(packages_required %in% installed.packages()[,"Package"])]

# Install any uninstalled packages
if(length(packages_uninstalled)) BiocManager::install(packages_uninstalled, ask = FALSE)
```

## - Loading packages

Here, we load the necessary packages for our analysis.

```{r message=FALSE, warning=FALSE}
library(hrbrthemes, quietly = TRUE)
library(fgsea, quietly = TRUE)
library(tidyr, quietly = TRUE)
library(escape, quietly = TRUE)
library(stringr, quietly = TRUE)
library(ggpubr, quietly = TRUE)
library(dplyr, quietly = TRUE)
library(ggplot2, quietly = TRUE)
```

## - Log Session Info

Finally, we log the session information for reproducibility.

```{r message=TRUE, warning=TRUE}
# Write the session information to a text file
writeLines(capture.output(sessionInfo()), file.path(script_folder, 'Fig_02_GH_Fig_S5_SessionInfo.txt'))

# Print the session information
sessionInfo()
```

# 2) Loading input files

```{r}
# Read Expression_matrix
Expression_matrix <- read.csv(
  file = paste(data_folder, '/RNASeq/InHouse/Counts/Processed_counts/Grouped/Tximport_TPM_mat.csv', sep = ''), 
  row.names = 'X'
)

# Replace 'X' in column names
colnames(Expression_matrix) <- str_replace_all(colnames(Expression_matrix), pattern = 'X', replacement = '')

# Read Genes_annotation_metadata
Genes_annotation_metadata <- read.csv(
  file = paste(data_folder, "/Resources/RNASeq/Annotation_files/GENCODE_v42.gene_metadata.csv", sep = '')
)
rownames(Genes_annotation_metadata) <- Genes_annotation_metadata$gene_id

# Read Samples_metadata
Samples_metadata <- read.csv(
  file = paste(data_folder, '/RNASeq/InHouse/Metadata/Samples_Metadata.csv', sep = ''), 
  row.names = 'Patient.ID'
)
```

# 3) Filtering expression matrix

```{r}
# Read unique genes
Unique_genes <- read.table(
  file = paste(data_folder, '/RNASeq/InHouse/Counts/Processed_counts/List_unique_genes.csv', sep = ''), 
  header = T, 
  sep = ','
)$unique_genes

# Filter Expression_matrix by unique genes
Expression_matrix <- Expression_matrix[Unique_genes, ]

# Replace row names with gene names
rownames(Expression_matrix) <- Genes_annotation_metadata[rownames(Expression_matrix),'gene_name']

# Convert to matrix
Expression_matrix <- as.matrix(Expression_matrix)

# Reorder columns based on Samples_metadata
Expression_matrix <- Expression_matrix[,rownames(Samples_metadata)]
```

# 4) Creating folders for Pathway Analyses

```{r}
dir.create(
  file.path(paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/BP-CMML_vs_Control', sep = '')), 
  showWarnings = FALSE
)
```

# 5) fGSEA Analyses 

## - Gene ranking

Here we read in a CSV file containing differentially expressed genes (DEGs) between the blast phase and control. We then calculate a rank for each gene based on its log2 fold change and p-value. The DEGs are arranged by their rank. Finally, we prepare a named numeric vector of ranks, removing any NA values.

```{r}
# Read DEGs
Blast_phase_vs_Control_DEGS <- read.csv(
  file = paste(data_folder, '/RNASeq/InHouse/DESeq2/DEGs_Comparisons/BP-CMML_vs_Control/DGE_BP-CMML_vs_Control_ALL_DEGs_LFC_Ashr.csv', sep = ''), 
  row.names = 'X'
)

# Calculate rank
Blast_phase_vs_Control_DEGS$Rank <- sign(Blast_phase_vs_Control_DEGS$log2FoldChange) * (-1 * log10(Blast_phase_vs_Control_DEGS$pvalue))

# Arrange by rank
Blast_phase_vs_Control_DEGS <- Blast_phase_vs_Control_DEGS %>% arrange(-Rank)

# Prepare ranks
Ranks <- Blast_phase_vs_Control_DEGS$Rank
names(Ranks) <- Blast_phase_vs_Control_DEGS$GeneSymbol
Ranks <- Ranks[!is.na(Ranks)]
```

## - Hallmark analysis

Here we read in a GMT file containing Hallmark pathways. We then perform a fast gene set enrichment analysis (fgsea) using the Hallmark pathways and the ranks of the DEGs. The results are arranged by the adjusted p-value.

We then write the full results to a CSV file. We also write two additional CSV files: one containing the pathways that are upregulated in the blast phase (BP-CMML) and one containing the pathways that are upregulated in the control.

```{r}
# Read Hallmark pathways
Hallmark_pathways_gmt_file <- paste(data_folder, '/Resources/RNASeq/Gene_Pathways/2022__GSEA_MSIGDB__h.all.v2022.1.Hs.symbols.gmt.txt', sep = '')
Hallmark_pathways <- gmtPathways(Hallmark_pathways_gmt_file)
str(head(Hallmark_pathways))

# Perform fgsea
fgseaRes_Hallmark <- fgsea(
  pathways = Hallmark_pathways, 
  stats = Ranks, 
  minSize=10, 
  maxSize = 500, 
  nPermSimple = 10000
)

# Arrange by padj
fgseaRes_Hallmark <- fgseaRes_Hallmark %>% arrange(padj)
fgseaRes_Hallmark$leadingEdge <- NULL

# Write results to csv
write.csv(
  x = fgseaRes_Hallmark, 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/BP-CMML_vs_Control/fGSEA_H_BP-CMML_vs_Control_all.csv', sep = '')
)

# Write upregulated in BP-CMML to csv
write.csv(
  x = fgseaRes_Hallmark %>% filter(NES > 0), 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/BP-CMML_vs_Control/fGSEA_H_BP-CMML_vs_Control__Up_in_BP-CMML.csv', sep = '')
)

# Write upregulated in Control to csv
write.csv(
  x = fgseaRes_Hallmark %>% filter(NES < 0), 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/BP-CMML_vs_Control/fGSEA_H_BP-CMML_vs_Control__Up_in_Control.csv', sep = '')
)
```

## - Canonical pathway analysis

Here we read in a GMT file containing Canonical pathways. We then perform a fast gene set enrichment analysis (fgsea) using the Canonical pathways and the ranks of the DEGs. The results are arranged by the adjusted p-value.

We then write the full results to a CSV file. We also write two additional CSV files: one containing the pathways that are upregulated in the blast phase (BP-CMML) and one containing the pathways that are upregulated in the control.

```{r}
# Read Canonical pathways
CanonicalPathways_gmt_file <- paste(data_folder, '/Resources/RNASeq/Gene_Pathways/2022__GSEA_MSIGDB__c2.cp.v2022.1.Hs.symbols.gmt.txt', sep = '')
CanonicalPathways <- gmtPathways(CanonicalPathways_gmt_file)
str(head(CanonicalPathways))

# Perform fgsea
fgseaRes_CP <- fgsea(
  pathways = CanonicalPathways, 
  stats = Ranks, 
  minSize=10, 
  maxSize = 500, 
  nPermSimple = 10000
)

# Arrange by padj
fgseaRes_CP <- fgseaRes_CP %>% arrange(padj)
fgseaRes_CP$leadingEdge <- NULL

# Write results to csv
write.csv(
  x = fgseaRes_CP, 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/BP-CMML_vs_Control/fGSEA_CP_BP-CMML_vs_Control_all.csv', sep = '')
)

# Write upregulated in BP-CMML to csv
write.csv(
  x = fgseaRes_CP %>% filter(NES > 0), 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/BP-CMML_vs_Control/fGSEA_CP_BP-CMML_vs_Control__Up_in_BP-CMML.csv', sep = '')
)

# Write upregulated in Control to csv
write.csv(
  x = fgseaRes_CP %>% filter(NES < 0), 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/BP-CMML_vs_Control/fGSEA_CP_BP-CMML_vs_Control__Up_in_Control.csv', sep = '')
)
```

## - GO BP pathway analysis

Here we read in a GMT file containing Gene Ontology Biological Process (GO_BP) pathways. We then perform a fast gene set enrichment analysis (fgsea) using the GO_BP pathways and the ranks of the DEGs. The results are arranged by the adjusted p-value.

We then write the full results to a CSV file. We also write two additional CSV files: one containing the pathways that are upregulated in the blast phase (BP-CMML) and one containing the pathways that are upregulated in the control.

```{r}
# Read GO_BP pathways
GO_BP_gmt_file <- paste(data_folder, '/Resources/RNASeq/Gene_Pathways/2022__GSEA_MSIGDB__c5.go.bp.v2022.1.Hs.symbols.gmt.txt', sep = '')
GOBP_pathways <- gmtPathways(GO_BP_gmt_file)
str(head(GOBP_pathways))

# Perform fgsea
fgseaRes_GO_BP <- fgsea(
  pathways = GOBP_pathways, 
  stats = Ranks, 
  minSize=10, 
  maxSize = 500, 
  nPermSimple = 10000
)

# Arrange by padj
fgseaRes_GO_BP <- fgseaRes_GO_BP %>% arrange(padj)
fgseaRes_GO_BP$leadingEdge <- NULL

# Write results to csv
write.csv(
  x = fgseaRes_GO_BP, 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/BP-CMML_vs_Control/fGSEA_GO_BP_BP-CMML_vs_Control_all.csv', sep = '')
)

# Write upregulated in BP-CMML to csv
write.csv(
  x = fgseaRes_GO_BP %>% filter(NES > 0), 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/BP-CMML_vs_Control/fGSEA_GO_BP_BP-CMML_vs_Control__Up_in_BP-CMML.csv', sep = '')
)

# Write upregulated in Control to csv
write.csv(
  x = fgseaRes_GO_BP %>% filter(NES < 0), 
  file = paste(data_folder, '/RNASeq/InHouse/Pathway_Analysis/BP-CMML_vs_Control/fGSEA_GO_BP_BP-CMML_vs_Control__Up_in_Control.csv', sep = '')
)
```

## - Pathway plotting

Here we combine gene set enrichment analysis results from different pathway sets, filter for specific pathways of interest, and create a bar plot of the Normalized Enrichment Score (NES) for each pathway, colored by cluster. The plot is then saved as a PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 2G', fig.pos='H', fig.height=4.0708661, fig.width=4.527559}
# Combine fgsea results into one data frame and sort by padj
Cumulative_fGSEAres <- rbind(fgseaRes_CP, fgseaRes_Hallmark, fgseaRes_GO_BP) %>% arrange(padj)

# Filter for positive NES and specific pathways
Cumulative_fGSEAres_posNES <- Cumulative_fGSEAres %>% 
  filter(NES > 0,
         pathway %in% c('HALLMARK_TNFA_SIGNALING_VIA_NFKB',
                        'REACTOME_NEUTROPHIL_DEGRANULATION',
                        'REACTOME_COMPLEMENT_CASCADE',
                        'GOBP_IMMUNE_RESPONSE_REGULATING_SIGNALING_PATHWAY',
                        'HALLMARK_INFLAMMATORY_RESPONSE',
                        'HALLMARK_IL6_JAK_STAT3_SIGNALING',
                        'HALLMARK_IL2_STAT5_SIGNALING',
                        'PID_AP1_PATHWAY'))

# Filter for negative NES and specific pathways
Cumulative_fGSEAres_negNES <- Cumulative_fGSEAres %>% 
  filter(NES < 0,
         pathway %in% c('HALLMARK_E2F_TARGETS',
                        'PID_MYC_ACTIV_PATHWAY',
                        'HALLMARK_MYC_TARGETS_V1',
                        'HALLMARK_MYC_TARGETS_V2',
                        'GOBP_DNA_REPLICATION',
                        'REACTOME_CELL_CYCLE_CHECKPOINTS',
                        'HALLMARK_G2M_CHECKPOINT',
                        'REACTOME_S_PHASE'))

# Combine positive and negative NES results
Cumulative_fGSEAres_top10 <- rbind(Cumulative_fGSEAres_posNES, Cumulative_fGSEAres_negNES)

# Assign cluster labels
Cumulative_fGSEAres_top10[1:8,'Cluster'] <- 'Blast phase'
Cumulative_fGSEAres_top10[9:16,'Cluster'] <- 'Control'

# Rename pathways for readability
Cumulative_fGSEAres_top10$pathway <- c('TNFA signaling via NFKB',
                                       'Neutrophil degranulation',
                                       "Complement cascade",
                                       'Immune response signaling',
                                       'Inflammatory response',
                                       'IL6/JAK2/STAT3 signaling',
                                       'IL2/STAT5 signaling',
                                       'AP1 pathway',
                                       'E2F targets',
                                       'MYC activation',
                                       'MYC targets V1',
                                       'MYC targets V2',
                                       'DNA replication',
                                       'Cell cycle checkpoints',
                                       'G2M checkpoint',
                                       'S phase')

# Create plot
plot <- ggplot(Cumulative_fGSEAres_top10, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=Cluster)) +
  coord_flip() +
  labs(x="", y="Normalized Enrichment\nScore (NES)",
       title="") + 
  theme_minimal(base_line_size = 0.3) + 
  scale_fill_manual(values = c('Blast phase'    = '#DD3429',
                               'Control'    = '#A4BECE')) +
  theme(legend.position = 'none',
        plot.title = element_text(face='bold', size =14, family = 'sans'),
        axis.title.y = element_text(size =12, family = 'sans'),
        axis.title.x = element_text(size =12, family = 'sans'))

# Print plot
print(plot)

# Save plot as PNG
png(filename = paste(figures_folder, '/Fig_02/Panel_G.png', sep = ''),
    res = 300, units = 'cm', height = 10.34, width  = 11.5)
print(plot)
invisible(dev.off())
```

## - Enrichment plots selected signatures

Here we read in a GMT file containing pathways related to quiescent vs dividing hematopoietic stem cells (HSCs). We then perform a fast gene set enrichment analysis (fgsea) using these pathways and the ranks of the DEGs.

We then plot the enrichment of the pathway with the smallest p-value and save the plot as a PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 2H', fig.pos='H', fig.height=2.625984, fig.width=3.46457}
# Define file path for Quiescence_HSC gmt file
Quiescence_HSC_gmt_file <- paste(data_folder, 
                        '/Resources/RNASeq/Gene_Pathways/2007__Graham_et_al__NormalQuiescent_vs_NormalDividing.gmt', 
                        sep = '')

# Read the gmt file into a list of pathways
Quiescence_HSC_pathways <- gmtPathways(Quiescence_HSC_gmt_file)

# Display the structure of the first few pathways
str(head(Quiescence_HSC_pathways))

# Perform fgsea on the pathways
fgseaRes_Quiescence_HSC_pathways <- fgsea(pathways = Quiescence_HSC_pathways, 
                           stats = Ranks, 
                           minSize=10, 
                           maxSize = 500, 
                           nPermSimple = 10000)

# Display the fgsea results
fgseaRes_Quiescence_HSC_pathways

# Plot the enrichment of the pathway with the smallest p-value
plot <- plotEnrichment(Quiescence_HSC_pathways[[head(fgseaRes_Quiescence_HSC_pathways[order(pval), ], 1)$pathway]],
               Ranks) + labs(title=head(fgseaRes_Quiescence_HSC_pathways[order(pval), ], 1)$pathway) + 
  ggtitle('UP in quiescence vs dividing HSC\nGraham et al. 2007') + 
  theme(plot.title = element_text(hjust = 0.5, face = 'bold', size = 13, family='sans')) +
  ylab('Enrichment score') + xlab('Rank')

# Print the plot
print(plot)

# Save the plot as a PNG file
png(filename = paste(figures_folder, '/Fig_02/Panel_H-1.png', sep = ''),
    res = 300, units = 'cm', height = 6.67, width  = 8.8)

# Print the plot again to save it to the file
print(plot)

# Close the PNG device
invisible(dev.off())
```

Here we read in a CSV file containing cell cycle-related gene signatures, perform a fast gene set enrichment analysis (fgsea) on these signatures. We also plot the enrichment of the G2M Phase signature in the results, saving the plot as PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 2H', fig.pos='H', fig.height=2.625984, fig.width=3.46457}
# Define file path for Roy2021 signatures csv file
Roy2021_signatures <- read.csv(paste(data_folder, 
                        '/Resources/RNASeq/Gene_Pathways/2021__Roy_et_al__cellcycle.csv', 
                        sep = ''))

# Convert the dataframe to a list and remove duplicate entries
Roy2021_signatures <- as.list(Roy2021_signatures)
Roy2021_signatures$G2M <- unique(Roy2021_signatures$G2M)[1:length(unique(Roy2021_signatures$G2M))-1]
Roy2021_signatures$S_Phase <- unique(Roy2021_signatures$S_Phase)[1:length(unique(Roy2021_signatures$S_Phase))-1]

# Perform fgsea on the signatures
Roy2021_signatures_results <- fgsea(pathways = Roy2021_signatures, 
                           stats = Ranks, 
                           minSize=10, 
                           maxSize = 500, 
                           nPermSimple = 10000)

# Display the fgsea results
Roy2021_signatures_results

# Plot the enrichment of the signature with the smallest p-value
plot <- plotEnrichment(Roy2021_signatures[[head(Roy2021_signatures_results[order(pval), ], 1)$pathway]],
               Ranks) + labs(title=head(Roy2021_signatures_results[order(pval), ], 1)$pathway) + 
  ggtitle('G2M phase\nTirosh et al. 2016') + 
  theme(plot.title = element_text(hjust = 0.5, face = 'bold', size = 13, family='sans')) +
  ylab('Enrichment score') + xlab('Rank')

# Print the plot
print(plot)

# Save the plot as a PNG file
png(filename = paste(figures_folder, '/Fig_02/Panel_H-2.png', sep = ''),
    res = 300, units = 'cm', height = 6.67, width  = 8.8)

# Print the plot again to save it to the file
print(plot)

# Close the PNG device
invisible(dev.off())
```

Here we create a plot of the enrichment of the S Phase signature in the cell cycle-related gene signatures results. We then save this plot as a PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 2H', fig.pos='H', fig.height=2.625984, fig.width=3.46457}
# Plot the enrichment of the third pathway in the Roy2021_signatures_results
plot <- plotEnrichment(Roy2021_signatures[[Roy2021_signatures_results[3,'pathway'][[1]]]],
               Ranks) + labs(title='') + 
  ggtitle('S phase\nTirosh et al. 2016') + 
  theme(plot.title = element_text(hjust = 0.5, face = 'bold', size = 13, family='sans')) +
  ylab('Enrichment score') + xlab('Rank')

# Print the plot
print(plot)

# Save the plot as a PNG file
png(filename = paste(figures_folder, '/Fig_02/Panel_H-3.png', sep = ''),
    res = 300, units = 'cm', height = 6.67, width  = 8.8)

# Print the plot again to save it to the file
print(plot)

# Close the PNG device
invisible(dev.off())
```

Here we create a plot of the enrichment of the quiescence signature in the cell cycle-related gene signatures results. We then save this plot as a PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 2H', fig.pos='H', fig.height=2.625984, fig.width=3.46457}
# Plot the enrichment of the second pathway in the Roy2021_signatures_results
plot <- plotEnrichment(Roy2021_signatures[[Roy2021_signatures_results[2,'pathway'][[1]]]],
               Ranks) + labs(title='') + 
  ggtitle('Quiescence\nGiustacchini et al. 2017') + 
  theme(plot.title = element_text(hjust = 0.5, face = 'bold', size = 13, family='sans')) +
  ylab('Enrichment score') + xlab('Rank')

# Print the plot
print(plot)

# Save the plot as a PNG file
png(filename = paste(figures_folder, '/Fig_02/Panel_H-4.png', sep = ''),
    res = 300, units = 'cm', height = 6.67, width  = 8.8)

# Print the plot again to save it to the file
print(plot)

# Close the PNG device
invisible(dev.off())
```

Here we read in a GMT file containing pathways related to senescence. We then perform a fast gene set enrichment analysis (fgsea) using these pathways and the ranks of the DEGs.

We then plot the enrichment of the pathway with the smallest p-value and save the plot as a PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure 2H', fig.pos='H', fig.height=2.625984, fig.width=3.46457}
# Define file path for Senescence gmt file
Senescence_gmt_file <- paste(data_folder, 
                        '/Resources/RNASeq/Gene_Pathways/2008__Fridman_and_Tainsky__UP_in_Senescencent.gmt', 
                        sep = '')

# Read the gmt file into a list of pathways
Senescence_pathways <- gmtPathways(Senescence_gmt_file)

# Display the structure of the first few pathways
str(head(Senescence_pathways))

# Perform fgsea on the pathways
fgseaRes_Senescence_pathways <- fgsea(pathways = Senescence_pathways, 
                           stats = Ranks, 
                           minSize=10, 
                           maxSize = 500, 
                           nPermSimple = 10000)

# Display the fgsea results
fgseaRes_Senescence_pathways

# Plot the enrichment of the pathway with the smallest p-value
plot <- plotEnrichment(Senescence_pathways[[head(fgseaRes_Senescence_pathways[order(pval), ], 1)$pathway]],
               Ranks) + labs(title=head(fgseaRes_Senescence_pathways[order(pval), ], 1)$pathway) + 
  ggtitle('Senescence UP\nFridman and Tainsky 2008') + 
  theme(plot.title = element_text(hjust = 0.5, face = 'bold', size = 13, family='sans')) +
  ylab('Enrichment score') + xlab('Rank')

# Print the plot
print(plot)

# Save the plot as a PNG file
png(filename = paste(figures_folder, '/Fig_02/Panel_H-5.png', sep = ''),
    res = 300, units = 'cm', height = 6.67, width  = 8.8)

# Print the plot again to save it to the file
print(plot)

# Close the PNG device
invisible(dev.off())
```

# 6) ssGSEA pathway enrichment

### - Fig S5

Here we define a list of pathways of interest from different sources (Hallmark, Canonical Pathways, and GO Biological Process). We then perform a single-sample Gene Set Enrichment Analysis (ssGSEA) on these pathways using the expression matrix.

We then create a boxplot showing the ssGSEA scores for each pathway, split between two conditions (BP-CMML and Control). We also add significance stars to the boxplot based on a t-test.

Finally, we save the plot as a PNG file.

```{r echo=TRUE, fig.align='center', fig.cap='Figure S6', fig.pos='H', fig.height=4.72441, fig.width=8.66142}
# Define pathways of interest from different sources
hallmark_pathways_of_interest <- list(
  'TNFA signaling via NFKB' = Hallmark_pathways$HALLMARK_TNFA_SIGNALING_VIA_NFKB,
  'Inflammatory response' = Hallmark_pathways$HALLMARK_INFLAMMATORY_RESPONSE,
  'IL6/JAK2/STAT3 signaling' = Hallmark_pathways$HALLMARK_IL6_JAK_STAT3_SIGNALING,
  'IL2/STAT5 signaling' = Hallmark_pathways$HALLMARK_IL2_STAT5_SIGNALING,
  'E2F targets' = Hallmark_pathways$HALLMARK_E2F_TARGETS, 
  'MYC targets V1' = Hallmark_pathways$HALLMARK_MYC_TARGETS_V1, 
  'MYC targets V2' = Hallmark_pathways$HALLMARK_MYC_TARGETS_V2, 
  'G2M checkpoint' = Hallmark_pathways$HALLMARK_G2M_CHECKPOINT
)

cp_pathways_of_interest <- list(
  'Neutrophil degranulation' = CanonicalPathways$REACTOME_NEUTROPHIL_DEGRANULATION,
  "Complement cascade" = CanonicalPathways$REACTOME_COMPLEMENT_CASCADE,
  'Cell cycle checkpoints' = CanonicalPathways$REACTOME_CELL_CYCLE_CHECKPOINTS,
  'S phase' = CanonicalPathways$REACTOME_S_PHASE,
  'AP1 pathway' = CanonicalPathways$PID_AP1_PATHWAY,
  'MYC activation' = CanonicalPathways$PID_MYC_ACTIV_PATHWAY
)

go_bp_pathways_of_interest <- list(
  'Immune response signaling' = GOBP_pathways$GOBP_IMMUNE_RESPONSE_REGULATING_SIGNALING_PATHWAY,
  'DNA replication' = GOBP_pathways$GOBP_DNA_REPLICATION
)

# Combine all pathways into a single collection
Gene_modules_collection <- c(hallmark_pathways_of_interest, cp_pathways_of_interest, go_bp_pathways_of_interest)

# Perform ssGSEA on the expression matrix using the defined pathways
Samples_scaled_results <- enrichIt(
  obj = as.matrix(Expression_matrix), 
  gene.sets = Gene_modules_collection, 
  cores = 12, 
  ssGSEA.norm = T,
  min.size = 5
)

# Add a 'Condition' column to the results
Samples_scaled_results[,'Condition'] <- 'BP-CMML'
Samples_scaled_results[c('929', '1126', '1138', '1157', '1162', '1170', '1207'),'Condition'] <- 'Control'

# Rename the columns of the results
colnames(Samples_scaled_results) <- c(
  "TNFA signaling via NFKB", 
  "IL6/JAK2/STAT3 signaling",
  "Inflammatory response",
  "IL2/STAT5/signaling",
  "E2F targets",
  "MYC targets V1", 
  "MYC targets V2", 
  "G2M checkpoint", 
  "Neutrophil degranulation",
  "Complement cascade",
  "Cell cycle checkpoints",
  "S phase",
  "AP1 pathway",
  "MYC activation",
  "Immune response signaling",
  "DNA replication",
  "Condition"
)

# Convert the data frame to a long format for plotting
long_data <- Samples_scaled_results %>%
  pivot_longer(cols = -Condition, names_to = "variable", values_to = "value")

# Define the order of the variables for the x-axis of the plot
variable_order <- c(
  "TNFA signaling via NFKB", 
  "Immune response signaling",
  "Neutrophil degranulation",
  "Inflammatory response", 
  "Complement cascade",
  "IL2/STAT5/signaling", 
  "IL6/JAK2/STAT3 signaling",
  "AP1 pathway",
  "S phase",
  "DNA replication",
  "Cell cycle checkpoints",
  "G2M checkpoint", 
  "MYC targets V2", 
  "MYC targets V1", 
  "MYC activation",
  "E2F targets"
)

# Convert the 'variable' column to a factor and specify the levels
long_data$variable <- factor(long_data$variable, levels = variable_order)

# Create a boxplot with significance stars, split between conditions
p <- ggplot(long_data, aes(x = variable, y = value, fill = Condition)) +
  geom_boxplot(position = position_dodge(0.9)) +
  geom_jitter(aes(color = Condition), position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.9), size = 1) +
  stat_compare_means(aes(group = Condition), label = "p.signif", method = "t.test") +
  scale_fill_manual(values = c('BP-CMML' = '#E35C53', 'Control' = '#B6CBD7')) +
  scale_color_manual(values = c('BP-CMML' = '#901900', 'Control' = '#5E6C7A')) +
  geom_vline(xintercept = which(variable_order == " "), linetype = "dashed") +
  theme_ipsum() +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1, size = 10),
    axis.title.x = element_blank(),
    axis.title.y = element_text(hjust = 0.5, vjust = 0.5, size = 12, face = 'bold')
  ) +
  labs(x = "Term", y = "Scaled\n ssGSEA score")

# Add arrows below the plot
p <- p + annotate("segment", x = which(variable_order == "AP1 pathway"), xend = which(variable_order == "TNFA signaling via NFKB"), y = min(long_data$value)-0.10, yend = min(long_data$value)-0.10, colour = "#DD3429", arrow = arrow(length = unit(0.3, "cm")))
p <- p + annotate("segment", x = which(variable_order == "S phase"), xend = which(variable_order == "E2F targets"), y = max(long_data$value)+0.10, yend = max(long_data$value)+0.10, colour = "#A4BECE", arrow = arrow(length = unit(0.3, "cm")))

# Print the plot
print(p)

# Save the plot as a PNG file
png(filename = paste(figures_folder, '/Supplemental/Fig_S5.png', sep = ''), res = 300, units = 'cm', height = 12, width  = 22)
print(p)
invisible(dev.off())
```

